Use statically-typed views for better performance

adlarkin opened the pull request:
Signed-off-by: Ashton Larkin <ashton@openrobotics.org>

# 🎉 New feature

Closes #711

## Summary
In an ongoing attempt to improve runtime performance, #711 points out that the runtime performance of `EntityComponentManager::Each` gets worse as more components are used. I've changed how views are implemented to improve the runtime of `EntityComponentManager::Each`. The main change here is that we now have a [statically-typed view](https://github.com/ignitionrobotics/ign-gazebo/blob/adlarkin/view_restructure/include/ignition/gazebo/detail/View.hh), which is defined by the list of component types that are required by the view. By doing this, a view can store an entity and pointers to the required component types in a [std::tuple](https://en.cppreference.com/w/cpp/utility/tuple), and then make use of [std::apply](https://en.cppreference.com/w/cpp/utility/apply) when we want to apply an entity and all of its relevant components to a callback in `EntityComponentManager::Each`.

Using `std::apply` allows for better runtime performance because we can apply the entity and associated components to a callback "all at once". This is quicker than the previous view approach, which involved individually looking up each component for an entity (see https://github.com/ignitionrobotics/ign-gazebo/issues/711#issuecomment-816890851).

The main drawback with the approach of a statically-typed view is the added software complexity/maintenance to the system. For example, during testing, I found that tuple creation can be costly - so, in order to avoid re-creating tuples whenever a component was being added/removed, I added an internal `ignore` flag to components to eliminate tuple re-creation. This `ignore` component flag is set to `true` whenever a user requests to remove an entity that was previously created, and is set to `false` whenever a user requests to add an entity that was previously removed. The `ignore` flag is hidden from the user and is used by the ECM and its views to determine which entities and components should be used in a particular `Each` call.

Another thing to note is that the [old ComponentStorage approach](https://github.com/ignitionrobotics/ign-gazebo/blob/ignition-gazebo5_5.0.0/include/ignition/gazebo/detail/ComponentStorageBase.hh) is no longer used. This new view approach does not require components of the same type to be packed sequentially in memory, and through various testing that I performed, I found that runtime performance with statically-typed views _**without**_ components stored sequentially in memory is still noticeably better than the runtime performance of the old views _**with**_ components stored sequentially in memory. So, hopefully this makes at least the `ComponentStorage` part of our system simpler to maintain :slightly_smiling_face: 

A side effect of removing the old `ComponentStorage` approach is that we also no longer need [ComponentId](https://github.com/ignitionrobotics/ign-gazebo/blob/ignition-gazebo5_5.0.0/include/ignition/gazebo/Types.hh#L78-L81), which means that we also no longer need [ComponentKey](https://github.com/ignitionrobotics/ign-gazebo/blob/ignition-gazebo5_5.0.0/include/ignition/gazebo/Types.hh#L88-L90). I tried to deprecate methods using `ComponentKey` wherever possible in order to follow the tick-tock model, but there were several methods that had to be deleted completely and replaced with something else (i.e., methods that returned `ComponentKey` or had `ComponentKey` as the only method parameter, which meant that there was no longer enough information provided to achieve the desired functionality of the method).

## Test it
If you'd like to compare the performance of running `Each` with the approach proposed in this PR to the existing view approach, you can run the [each benchmark](https://github.com/ignitionrobotics/ign-gazebo/tree/ign-gazebo5/test/benchmark). I have already done this on my system, and have included the results below (I ran a comparison against https://github.com/ignitionrobotics/ign-gazebo/commit/8f5103bfac4e748cbc0bb58d36991feee6554196):

[Each benchmark test](https://gist.github.com/adlarkin/485bca816eb4ab59201b35d417c77757#file-comparison-txt)

As seen from the results above, all of the `EachCache` tests are significantly faster with the statically-typed view approach - in fact, if we look at the very last test in that file (`Each10ComponentCache/1000` - so, 1000 entities with 10 components each), we go from 150ms to 2ms :exploding_head: It's also worth noting that we don't see a decrease in `EachNoCache` performance either, which makes sense since views are not used for `EachNoCache`.

---

Another way to test the changes here is to run a world with [3k shapes](https://gist.github.com/adlarkin/c0a97f50bfaa04131ec7275eaa3f7c0f). Here are the RTF numbers I'm seeing from testing 3k simple shapes, using both GUI and headless simulation. I used TPE for the physics engine. I also echoed the `/world/shapes/dynamic_pose/info` topic to see how RTF was impacted using these changes (see #743).

original ECM (`main` at https://github.com/ignitionrobotics/ign-gazebo/commit/ca10c7744f6d2e7146b95f0b05e1e98acdd02abf):
* static, no gui: 42%
    * echoing the `/world/shapes/dynamic_pose/info` topic: 8.3%
* static, gui: bounces around between 11-15%
    * echoing the `/world/shapes/dynamic_pose/info` topic: 4%
* non-static, no gui: 8.5%
    * echoing the `/world/shapes/dynamic_pose/info` topic: 3.5%
* non-static, gui: 4%
    * echoing the `/world/shapes/dynamic_pose/info` topic: 2.3%

new ECM (this branch):
* static, no gui: 50%
    * echoing the `/world/shapes/dynamic_pose/info` topic: 35% (this is a huge increase compared to 8.3% :exploding_head:)
* static, gui: bounces around 30-35%
    * echoing the `/world/shapes/dynamic_pose/info` topic: 23% (again, a huge increase compared to 4% :exploding_head:)
* non-static, no gui: 8.6%
    * echoing the `/world/shapes/dynamic_pose/info` topic: 5%
* non-static, gui: 6%
    * echoing the `/world/shapes/dynamic_pose/info` topic: 3.75%

As the numbers show, there's a performance improvement for almost every single test case (there's not much of an improvement for non-static, no GUI). For use cases where plugins are used that make calls to `Each` with a large number of components, I suspect that there will be even greater performance improvements.

It was also pointed out in https://github.com/ignitionrobotics/ign-gazebo/pull/856#issuecomment-859039148 that GUI startup time for 3k shapes with the changes in this PR is only a few seconds (it used to take up to a few minutes). Perhaps this is because of some `each` calls that are used when initializing the GUI?

Another thing to note is that when using the changes in this PR to run 3k shape simulations with a GUI, loading time is a lot faster. In the tests I ran, I found that these new changes allowed the GUI to load and run in a few seconds, while the old ECM/View implementation took a few minutes to load (thanks @iche033 for making this discovery!).

---

One other scenario that could be worth testing is a benchmark of the time it takes to add/remove components frequently. I did some informal testing locally with the new approach and found that removing components with the statically-typed views and an internal component `ignore` flag is faster than the current ECM/view approach, but perhaps it would be worth adding a benchmark test for this. This PR is already very large, so I decided to hold off on adding a benchmark test for now, but if anyone would be interested in me adding this in, let me know - I can also make another PR that adds this benchmark test if that is more preferable :slightly_smiling_face: 

## Takeaways/Next steps

While it's clear that this PR has improved the runtime performance of `EntityComponentManager::Each`, there's not much of an improvement in RTF for the headless tests. While #793 should help improve the RTF for running the GUI, we will need to figure out why we still aren't seeing higher RTF for headless simulation (theoretically, headless RTF for 3k static shapes should be 100%).

I would also be interested in figuring out how to generate better testing scenarios/evaluation metrics. Perhaps 3k simple shapes isn't a great way to benchmark performance - we aren't running any plugins other than the physics system, and aren't doing things like adding/removing components. Maybe an external user will try these changes in a context that differs from the 3k simple shapes example and find a nice performance improvement! Considering that most users use various plugins and may or may not add/remove components frequently, I think it'd be worth taking some time to figure out how to benchmark use cases that are frequent among Ignition users. That way, we can get a better feel for whether our changes are being noticed by the community or not.

Also, it would be nice to add some more features/functionality to the `EntityComponentManager` that make use of the new statically-typed views. For example, perhaps we can use iterators (as done in [EnTT](https://github.com/skypjack/entt)) for use cases where a lot of components are used as "filters", and we only need specific components of entities that pass all component filter checks. So, if a user wanted all entities that are static models, but only need the pose of these models, then maybe they could do something like this:
```c++
auto entityGroup = ecm.FilterEntities<components::Static, components::Model, components::Pose>();

for (auto entity : entityGroup)
{
  auto pose = entity.Component<components::Pose>();

  // do something with pose here
}
```

Related to the idea of providing functionality like iterators as described above, we will eventually need to address #628 and #805.

## Checklist
- [x] Signed all commits for DCO
- [x] Added tests
- [x] Added example and/or tutorial
- [x] Updated documentation (as needed)
- [x] Updated migration guide (as needed)
- [x] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [x] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

adlarkin commented on test/integration/components.cc:
```diff
@@ -966,7 +966,7 @@ TEST_F(ComponentsTest, LogicalMicrophone)
   EXPECT_EQ("0 0.5", ostr.str());
 
-  std::istringstream istr;
+  std::istringstream istr(ostr.str());
```
This change was made temporarily to get this test to pass - the real fix is in https://github.com/ignitionrobotics/ign-gazebo/pull/848. Once that PR is merged and forward-ported, I can revert this change.

adlarkin commented on include/ignition/gazebo/Types.hh:
```diff
@@ -78,6 +78,8 @@ namespace ignition
     /// \brief A unique identifier for a component instance. The uniqueness
     /// of a ComponentId is scoped to the component's type.
     /// \sa ComponentKey.
+    // TODO(anyone) remove this in ign-gazebo7. This is no longer used in
+    // ign-gazebo6
     using ComponentId = int;
```
I left a TODO note to remove this because I don't think it's possible to mark a type alias as deprecated, right?

adlarkin commented on include/ignition/gazebo/Types.hh:
```diff
@@ -87,6 +89,8 @@ namespace ignition
 
     /// \brief A key that uniquely identifies, at the global scope, a component
     /// instance
+    // TODO(anyone) remove this in ign-gazebo7. This is no longer used in
+    // ign-gazebo6
     using ComponentKey = std::pair<ComponentTypeId, ComponentId>;
```
I left a TODO note to remove this because I don't think it's possible to mark a type alias as deprecated, right?

adlarkin commented on src/EntityComponentManager_TEST.cc:
```diff
@@ -107,149 +107,24 @@ class EntityComponentManagerFixture : public ::testing::TestWithParam<int>
 };
 
-/////////////////////////////////////////////////
-TEST_P(EntityComponentManagerFixture, AdjacentMemorySingleComponentType)
```
I removed tests related to adjacent/sequential memory since the old `ComponentStorage` approach is no longer used.

adlarkin reviewed (commented):
I've got a few questions/notes for my reviewers.

iche033 commented:
Do you notice that when launching `shapes_population.sdf` with GUI, the scene is now loaded faster? It used to take 2 mins on my machine and now it's only just a few seconds.

I am testing with 2 workspaces, one on this branch and the other on `main`, and both built from source. My RTFs from running the shapes population world with the tpe plugin:

new ECM (this branch):
* static, no gui: 35%
* static, gui: 16%
* non-static, no gui: 10%
* non-static, gui: 5%

original ECM (`main`)
* static, no gui: 26%
* static, gui: 6%
* non-static, no gui: 9%
* non-static, gui:  3.5%

So looks like it made a difference on my machine. Could you try comparing against fortress built from source and see if you also see these improvements?

adlarkin commented:
> Do you notice that when launching shapes_population.sdf with GUI, the scene is now loaded faster? It used to take 2 mins on my machine and now it's only just a few seconds.

I never ran this with the GUI the first time around, but now that you have made the point, I went ahead and tested it myself. Yes, the scene loads a lot faster for me now with these changes. I'm also seeing loading times decrease from a few minutes to a few seconds. Thanks for pointing this out!

> Could you try comparing against fortress built from source and see if you also see these improvements?

I went ahead an did a comparison with source builds and have updated the PR description with my results. I do notice some improvements now, especially when echoing the `/world/shapes/dynamic_pose/info` topic :sunglasses: (see #743)

iche033 reviewed (commented):
Left a few minor optimization comments in code.

I'm also randomly testing examples worlds. So far I ran into one crash. To reproduce:

```
ign gazebo -v 4 -r --levels levels.sdf
```

Right click and remove the red vehicle -> segfault

adlarkin commented on src/EntityComponentManager.cc:
```diff
@@ -342,48 +353,70 @@ void EntityComponentManager::ProcessRemoveEntityRequests()
 bool EntityComponentManager::EntityHasComponent(const Entity _entity,
     const ComponentKey &_key) const
 {
+  ignwarn << "This method is deprecated. Please use "
```
https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/ee9b6c50adef4fb2641fa630dbd1a3d7b453f0e0

adlarkin commented on src/EntityComponentManager.cc:
```diff
@@ -342,48 +353,70 @@ void EntityComponentManager::ProcessRemoveEntityRequests()
+bool EntityComponentManager::RemoveComponent(
+    const Entity _entity, const ComponentKey &_key)
+{
+  ignwarn << "This method is deprecated. Please use the RemoveComponent method "
```
https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/ee9b6c50adef4fb2641fa630dbd1a3d7b453f0e0

adlarkin commented on src/EntityComponentManager.cc:
```diff
@@ -152,10 +159,11 @@ class ignition::gazebo::EntityComponentManagerPrivate
+  /// \brief Unordered map of removed components. The key is the entity to
+  /// which belongs the component, and the value is a set of the component types
+  /// being removed.
+  std::unordered_map<Entity, std::unordered_set<ComponentTypeId>>
```
https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/ee9b6c50adef4fb2641fa630dbd1a3d7b453f0e0

adlarkin commented on include/ignition/gazebo/detail/EntityComponentManager.hh:
```diff
@@ -167,23 +166,26 @@ bool EntityComponentManager::SetComponentData(const Entity _entity,
 {
-  return static_cast<ComponentTypeT *>(
-      this->First(ComponentTypeT::typeId));
+  ignwarn << "This method is now deprecated and will always return nullptr.\n";
```
https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/ee9b6c50adef4fb2641fa630dbd1a3d7b453f0e0

adlarkin commented on include/ignition/gazebo/detail/EntityComponentManager.hh:
```diff
@@ -167,23 +166,26 @@ bool EntityComponentManager::SetComponentData(const Entity _entity,
 {
-  return static_cast<const ComponentTypeT *>(
-      this->First(ComponentTypeT::typeId));
+  ignwarn << "This method is now deprecated and will always return nullptr.\n";
```
https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/ee9b6c50adef4fb2641fa630dbd1a3d7b453f0e0

adlarkin commented on src/EntityComponentManager.cc:
```diff
@@ -342,48 +353,70 @@ void EntityComponentManager::ProcessRemoveEntityRequests()
+  {
+    periodicIter->second.erase(_entity);
+    if (periodicIter->second.empty())
+      this->dataPtr->periodicChangedComponents.erase(_typeId);
```
https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/39b499e0eef43623a21c6ec277411c75140b8a98

adlarkin commented on src/EntityComponentManager.cc:
```diff
@@ -342,48 +353,70 @@ void EntityComponentManager::ProcessRemoveEntityRequests()
+  {
+    oneTimeIter->second.erase(_entity);
+    if (oneTimeIter->second.empty())
+      this->dataPtr->oneTimeChangedComponents.erase(_typeId);
```
https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/39b499e0eef43623a21c6ec277411c75140b8a98

adlarkin commented on src/ComponentStorage.cc:
```diff
@@ -0,0 +1,170 @@
+{
+  if (this->entityComponents.find(_entity) != this->entityComponents.end())
+    return false;
+  this->entityComponents[_entity];
```
https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/39b499e0eef43623a21c6ec277411c75140b8a98

adlarkin commented on src/ComponentStorage.cc:
```diff
@@ -0,0 +1,170 @@
+  // make sure the entity exists
+  auto typeMapIter = this->componentTypeIndex.find(_entity);
+  auto entityCompIter = this->entityComponents.find(_entity);
+  if (typeMapIter == this->componentTypeIndex.end() ||
```
https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/39b499e0eef43623a21c6ec277411c75140b8a98

adlarkin commented on include/ignition/gazebo/detail/BaseView.hh:
```diff
@@ -0,0 +1,199 @@
+  /// \brief See if an entity is marked as an entity to be added to the view
+  /// \param[in] _entity The entity
+  /// \return true if _entity is to be added to the view, false otherwise
+  public: bool HasEntityToAdd(const Entity _entity) const;
```
https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/39b499e0eef43623a21c6ec277411c75140b8a98

adlarkin commented on include/ignition/gazebo/detail/BaseView.hh:
```diff
@@ -0,0 +1,199 @@
+  /// \return True if _entity isn't already a part of the view and was marked as
+  /// an entity to be added. False otherwise
+  /// \sa HasEntity, HasEntityToAdd
+  public: bool MarkEntityToAdd(const Entity _entity, const bool _new = false);
```
IsEntityMarkedForAddition

adlarkin commented:
I've addressed all review feedback, except for updating the migration guide with things that had to be deleted instead of deprecated (I'll do this once we know no other changes have to be made).

> So far I ran into one crash which I'm not able to reproduce on the main branch. To reproduce:
> ign gazebo -v 4 -r --levels levels.sdf
> Right click and remove the red vehicle -> segfault

I tested this locally as well, and also noticed a segfault. I will debug this and update the PR once I've found a fix.

chapulina reviewed (commented):
Just did a first pass. There's a lot of changes, so please allow me a few passes until I get the hang of it.

I have some suggestions to allow us to tick-tock most of the API and ease migration in https://github.com/ignitionrobotics/ign-gazebo/pull/873.

Meanwhile, can you add some unit tests for the new classes / files? i.e. `BaseView`, `ComponentStorage`

adlarkin commented on src/EntityComponentManager.cc:
```diff
@@ -78,20 +76,28 @@ class ignition::gazebo::EntityComponentManagerPrivate
-          std::unique_ptr<ComponentStorageBase>> components;
+  /// \brief A class that stores all components and maps entities to their
+  /// component types
+  public: detail::ComponentStorage componentStorage;
```
They are pretty similar, and do indeed have some overlap. When I was originally working on this PR, I actually wondered to myself if `entityComponents` is even needed. I think that the answer to this question is no, but for the time being, I decided to leave `entityComponents` as-is since this PR is already quite large.

What if I leave a `TODO` note and/or ticket an issue about this, and then work on replacing `entityComponents` with `componentStorage` in a separate PR after this is merged? That way, we can avoid adding any more changes to this PR.

chapulina replied:
My concern is that with the current implementation, it's not clear to me what's the source of truth when it comes to entities and their components. 

> replacing entityComponents with componentStorage in a separate PR after this is merged? 

It sounds like we may need to keep `entityComponents` and `componentStorage` in sync, and this may not be what's currently happening. This may be resulting in subtle inconsistencies that our tests may not be picking up. Since `componentStorage` is being introduced in this PR, I'd like to understand whether it's needed before merging it in. If it's a replacement for `entityComponents`, then we should straight out remove that. If it serves a different purpose, I'd like to understand how they're related to each other so we can cover our bases. I'm worried that this may be more pervasive than a simple implementation detail that can be fixed on a future PR.

adlarkin replied:
`componentStorage` was intended to replace the old component storage, which stored components sequentially in memory: https://github.com/ignitionrobotics/ign-gazebo/blob/9fd8618436e49691fd0cbdaf5bf054ee9384b857/src/EntityComponentManager.cc#L81-L84

> My concern is that with the current implementation, it's not clear to me what's the source of truth when it comes to entities and their components.

I think that this is a little confusing because of how the component storage class was changed, so let me try to explain how component storage worked before this PR vs how it works now (or at least, explain them as I understand it :slightly_smiling_face:):

Before this PR, we stored components in the [`ComponentStorageBase`](https://github.com/ignitionrobotics/ign-gazebo/blob/ignition-gazebo5_5.0.0/include/ignition/gazebo/detail/ComponentStorageBase.hh) class, which stored components of a particular type sequentially in memory. A setback with this approach is that for a particular `ComponentStorageBase` instance, we know the type of component being stored, but we don't have any information about which entity owns a particular component instance in the storage. In order to solve this issue, we use `entityComponents` to store the connecting/missing information: https://github.com/ignitionrobotics/ign-gazebo/blob/9fd8618436e49691fd0cbdaf5bf054ee9384b857/src/EntityComponentManager.cc#L116-L120

As you can see, for a given entity, we keep track of the component types it has (`ComponentTypeId`) and where to index the memory in the component type's `CompoentStorageBase` instance (`ComponentId`) in order to get the component instance of a particular type that belongs to this entity. Here's an example of how `entityComponents` is used to get a component of a particular type that belongs to an entity: https://github.com/ignitionrobotics/ign-gazebo/blob/9fd8618436e49691fd0cbdaf5bf054ee9384b857/src/EntityComponentManager.cc#L606-L623

So, with the previous implementation, we needed `entityComponents` because the component storage alone did not give us enough information when it came to mapping entities to their components.

Now, let's look at the implementation proposed in this PR. The new `ComponentStorage` class maps an entity to all of its components directly, because we are no longer storing component instances sequentially in memory: https://github.com/ignitionrobotics/ign-gazebo/blob/046210aa735681d83ffbbd0d005c38b33541053d/include/ignition/gazebo/detail/ComponentStorage.hh#L128-L145

This also means that we no longer need a separate component storage instance for each component type - now, we only need once component storage instance, since this single instance will manage all components of every type: https://github.com/ignitionrobotics/ign-gazebo/blob/046210aa735681d83ffbbd0d005c38b33541053d/src/EntityComponentManager.cc#L79-L81

Because of this change, we don't need `entityComponents` anymore - if we want to get a component of a particular type that belongs to a specific entity, we don't need another container that connects entities to their component instances. Instead, we can call this method on the new `ComponentStorage` class: https://github.com/ignitionrobotics/ign-gazebo/blob/046210aa735681d83ffbbd0d005c38b33541053d/include/ignition/gazebo/detail/ComponentStorage.hh#L111-L121

So, as you can see, with these new changes, `entityComponents` is essentially a "convenient container" which provides us with a simple way to check what types of components an entity has: https://github.com/ignitionrobotics/ign-gazebo/blob/046210aa735681d83ffbbd0d005c38b33541053d/src/EntityComponentManager.cc#L122-L126

That being said though, I really don't think `entityComponents` is needed anymore, because the original intent/usage of `entityComponents` (connect entities to their component instances in component storage) is now self-contained in the new `ComponentStorage` class.

So, you're probably wondering, "if we don't need `entityComponents` anymore, then why hasn't it been removed?" I haven't removed it yet for a few reasons:
1. I only wanted to change what was absolutely necessary when creating this PR, because I already knew that in order to meet the bare minimum of getting code to compile, a lot of changes would have to be made. So, I figured that changes like these could wait until it came time for review in order to make it easier for the PR reviewers.
2. When I was originally working on this PR, I was _fairly confident_ that we didn't need `entityComponents` anymore, but I wasn't 100% sure. However, now that I have given you this explanation, I am confident that we no longer need `entityComponents` thanks to the new [`ComponentStorage`](https://github.com/ignitionrobotics/ign-gazebo/blob/046210aa735681d83ffbbd0d005c38b33541053d/include/ignition/gazebo/detail/ComponentStorage.hh) class :grin: 

So, to recap/summarize: in the old implementation, both `ComponentStorageBase` and `entityComponents` were the source of truth - both were needed in order to store components and map which components belonged to which entity. However, with the new implementation, `ComponentStorage` is the source of truth, and `entityComponents` is no longer needed because the new `ComponentStorage` class provides the mapping of entities to their components.

> It sounds like we may need to keep `entityComponents` and `componentStorage` in sync

As a follow-up from my (long) explanation above, if we decide to keep `entityComponents`, then yes, I agree :+1: however, I don't believe that `entityComponents` is needed anymore.

> I'd like to understand how they're related to each other so we can cover our bases

Hopefully the explanation I gave above helps you understand how `ComponentStorage` relates to `EntityComponents` :sweat_smile:

I'll just add one more note regarding "covering our bases": if we do decide to remove `entityComponents`, we'll need to be sure that we still handle updating `entityComponentsDirty` correctly: https://github.com/ignitionrobotics/ign-gazebo/blob/39b499e0eef43623a21c6ec277411c75140b8a98/src/EntityComponentManager.cc#L117-L135

This shouldn't be too difficult, but I just wanted to mention it here as a reminder.

chapulina replied:
Thank you for the explanation, very clear! Good to hear we don't need `entityComponents` anymore and can reduce some of the complexity.

Have you started working on removing that on a separate branch or are you waiting to see if the approach here is accepted before working on that?

adlarkin replied:
> Have you started working on removing `entityComponents` on a separate branch or are you waiting to see if the approach here is accepted before working on that?

I have not started removing this yet - my plan was to wait for everything else to be reviewed/approved first before making this change, since this PR is already quite large. I can create a new branch to work on this though if you'd like - just let me know!

> The new component storage class offers APIs such as adding / removing entities, creating and fetching components... This is pretty much what the `EntityComponentManager` does. Why do we need another class? Couldn't these maps just be directly inside `EntityComponentManager`?

I guess that we could store these maps in the `EntityComponentManager`, but I'm not a fan of this approach for a few reasons:
1. The `EntityComponentManager` class is already quite large, and adding the functionality from `ComponentStorage` to `EntityComponentManager` directly would make `EntityComponentManager` even more complex. There are also a few lower-level things going on here with a component's `removed` flag, which I think justifies having a new class for the functionality of mapping entities to their components
2. Breaking things down into multiple classes makes it easier for readers of the code to find what they're looking for (and it makes it easier for us as maintainers, too :slightly_smiling_face:)
3. Having a separate class for `ComponentStorage` means we can test this class directly. If the functionality was put directly into `EntityComponentManager`, it would be harder to test it. While we could make the API in `ComponentStorage` part of the public API for `EntityComponentManager` so that it's easier to test, I don't think this would make much sense since the `ComponentStorage` API is "low level" and shouldn't be used by our users directly (they should revert to the "higher level" API provided by `EntityComponentManager`).

> I'm also not clear why the storage class is currently under detail instead of src. It isn't templated or used by any templates.

You're right :slightly_smiling_face: when I was originally working on this class, I had a templated method that was later removed because it wasn't needed. I have moved this over to the regular `include` and `src` directories in https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/4a90e86a00445802f2434cda718b1cd17b90ca78

adlarkin replied:
marking this as resolved thanks to #943

adlarkin commented on include/ignition/gazebo/detail/BaseView.hh:
```diff
@@ -0,0 +1,199 @@
+  protected: std::unordered_map<Entity, bool> toAddEntities;
+
+  /// \brief The component types in the view
+  protected: std::set<ComponentTypeId> componentTypes;
```
This is used with the `EntityMatches` method (notice how the `_types` parameter is a `std::set`): https://github.com/ignitionrobotics/ign-gazebo/blob/39b499e0eef43623a21c6ec277411c75140b8a98/include/ignition/gazebo/EntityComponentManager.hh#L143-L148

Here's an example use case: https://github.com/ignitionrobotics/ign-gazebo/blob/39b499e0eef43623a21c6ec277411c75140b8a98/include/ignition/gazebo/detail/EntityComponentManager.hh#L504-L506

I could change this to be `unordered`, but that would also mean another `EntityMatches` method with a `std::unordered_set` parameter type needs to be created, and the existing `EntityMatches` method that takes a `std::set` should be deprecated. If you want me to make this change, I've got no problem with it, but I figured that it may not be worth going through another deprecation because I don't think the performance gains of using a `std::unoredered_set` for `componentTypes` will be very noticeable - `componentTypes` is populated when the view is created, and is never modified after that: https://github.com/ignitionrobotics/ign-gazebo/blob/39b499e0eef43623a21c6ec277411c75140b8a98/include/ignition/gazebo/detail/View.hh#L49-L53

So, considering the fact that 1) we only need to insert elements into `componentTypes` once, 2) the elements are a primitive type (`uint64_t`), which reduces complexity for insertion comparison, and 3) it's unlikely that the number of elements in `componentTypes` will be very large (most views consist of less than 10 types), I wasn't sure if adding a new `EntityMatches` method (and deprecating the old one) was worth it for _what I believe to be_ a very minor optimization (it'd be a `O(1)` operation instead of `O(logn)` for a few insertions that only happens once).

What do you think? Should we change this to `std::unordered_set` and add a new `EntityMatches` method? Another option would be to not deprecate the old `EntityMatches` method and keep the two methods with different input parameter types, but I don't see the value in keeping two different method signatures here.

adlarkin commented on src/EntityComponentManager_TEST.cc:
```diff
@@ -1418,6 +1197,55 @@ TEST_P(EntityComponentManagerFixture, EachGetsNewOldRemove)
+  comp = manager.Component<IntComponent>(e1);
+  ASSERT_NE(nullptr, comp);
+  EXPECT_EQ(789, comp->Data());
+  manager.RemoveComponent(e1, IntComponent::typeId);
```
This specific test case is not supposed to call `eachCount` before removing again, because I am testing that the corresponding view's [`toAddEntities` queue](https://github.com/ignitionrobotics/ign-gazebo/blob/39b499e0eef43623a21c6ec277411c75140b8a98/include/ignition/gazebo/detail/BaseView.hh#L188-L190) is being updated properly: https://github.com/ignitionrobotics/ign-gazebo/blob/f8204deb4372993519e856ad020e2e59a5e3a158/src/EntityComponentManager_TEST.cc#L1330-L1334

Since a lot of this code is probably still pretty new to you, here's what is happening in this specific test code block:
1. call `CreateComponent` on an entity with a component type of `IntComponent`, which makes this entity valid for the `IntComponent` view and adds it to the view's `toAddEntities` queue (notice the `view->MarkEntityToAdd` call): https://github.com/ignitionrobotics/ign-gazebo/blob/39b499e0eef43623a21c6ec277411c75140b8a98/src/EntityComponentManager.cc#L603-L611
2. remove the component from the entity, which makes this entity no longer valid for the view, so the entity should be removed from the view's `toAddEntities` queue (notice the `NotifyComponentRemoval` call): https://github.com/ignitionrobotics/ign-gazebo/blob/39b499e0eef43623a21c6ec277411c75140b8a98/src/EntityComponentManager.cc#L381-L388
3. call `eachCount`, which internally calls `ecm.Each`. When `ecm.Each` is called, the view's `toAddEntities` queue should be empty, so the view shouldn't have any entities in it because it was initially empty and has no entities to be added to it: https://github.com/ignitionrobotics/ign-gazebo/blob/39b499e0eef43623a21c6ec277411c75140b8a98/include/ignition/gazebo/detail/EntityComponentManager.hh#L483-L492

So, as you can see, I am intentionally calling `eachCount` after `RemoveComponent` here because I want to make sure that the view empties its `toAddEntities` queue when needed. If I called `eachCount` before `RemoveComponent`, the view would clear the `toAddEntities` queue after adding all entities from the queue as a part of the `Each` call, and I wouldn't have a way to test this.

Does that make sense?

adlarkin commented on src/EntityComponentManager_TEST.cc:
```diff
@@ -1418,6 +1197,55 @@ TEST_P(EntityComponentManagerFixture, EachGetsNewOldRemove)
+  EXPECT_EQ(nullptr, comp);
+
+  // add a component
+  manager.CreateComponent<IntComponent>(e1, IntComponent(123));
```
I added `nullptr` checks and `true` checks for all `CreateComponent` and `RemoveComponent` calls in https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/f8204deb4372993519e856ad020e2e59a5e3a158

adlarkin commented on include/ignition/gazebo/components/Component.hh:
```diff
@@ -275,6 +280,21 @@ namespace components
+    /// achieving removal (and re-addition) behavior for a component that
+    /// already exists without actually removing the component from the ECM
+    /// (using an internal flag instead of actually removing and re-adding the
+    /// component yields better runtime performance). This flag is to be used by
```
I added a note about this in https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/cd635dd3d93a358ea179e9bdd5378682e6e57b1f

adlarkin commented on include/ignition/gazebo/components/Component.hh:
```diff
@@ -275,6 +280,21 @@ namespace components
+    /// (using an internal flag instead of actually removing and re-adding the
+    /// component yields better runtime performance). This flag is to be used by
+    /// the ECM only, and should not change how users interact with components.
+    private: bool ignore{false};
```
`removed` works fine - in fact, I think it's actually better than `ignore`. This flag is meant for marking removal only, so I made the change in https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/cd635dd3d93a358ea179e9bdd5378682e6e57b1f

adlarkin commented on include/ignition/gazebo/detail/BaseView.hh:
```diff
@@ -0,0 +1,199 @@
+
+  /// \brief Get all of the entities to be added to the view
+  /// \return The entities to be added to the view
+  public: const std::unordered_map<Entity, bool> &ToAddEntities() const;
```
I added documentation for this in https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/52f8c7eedbd44629b97860dcf804207d49414788. Let me know if it's still unclear in any way!

adlarkin commented on include/ignition/gazebo/detail/BaseView.hh:
```diff
@@ -0,0 +1,199 @@
+  public: void ClearToAddEntities();
+
+  /// \brief Destructor
+  public: virtual ~BaseView()
```
https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/52f8c7eedbd44629b97860dcf804207d49414788

adlarkin commented on include/ignition/gazebo/detail/ComponentStorage.hh:
```diff
@@ -0,0 +1,151 @@
+  FAILED_ADDITION = 3
+};
+
+class IGNITION_GAZEBO_HIDDEN ComponentStorage
```
Documentation has been added in https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/52f8c7eedbd44629b97860dcf804207d49414788

> Also, since it serves a different purpose from the old `ComponentStorage`

This class actually doesn't serve a different purpose from the old `ComponentStorage` - both this class and the old storage class are meant to store components that have been created. The main difference between the two classes is that the new class does not store components sequentially in memory, while the old class does. This change/difference allows for the new class to keep an internal mapping of which component an entity belongs to, which was not possible with the old class (see https://github.com/ignitionrobotics/ign-gazebo/pull/856#discussion_r659100121 for more information)

adlarkin replied:
Since we removed this class in #943, I'm marking this as resolved

chapulina commented on include/ignition/gazebo/detail/View.hh:
```diff
@@ -34,110 +36,250 @@ namespace gazebo
+  }
+
+  /// \brief Documentation inherited
+  public: bool HasCachedComponentData(const Entity _entity) const override
```
Nit: can we separate the function declarations from the implementations? See the old `detail/ComponentStorageBase.hh` file for an example. That makes it easier to see all the API.

adlarkin replied:
Yep :+1: see https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/ea7d9e9b281e30d03ca89d3b22879f56fe53194b

adlarkin commented on include/ignition/gazebo/detail/ComponentStorage.hh:
```diff
@@ -0,0 +1,156 @@
+{
+  /// \brief Clear all of the entity and component data that has been tracked.
+  /// This "resets" the storage to its initial, empty state.
+  public: void Reset();
```
I replaced `Reset()` with a destroy/reconstruct in https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/782844ac8f419721b82854e419b5ee9f407bc720

adlarkin commented:
I believe that all feedback has been addressed, and the migration guide has also been updated in https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/f0a0fe85029fe448b1c12ca522db8ce6f67dc71d (please let me know if the migration guide is missing anything). This is ready for another round of review!

chapulina commented on include/ignition/gazebo/components/Factory.hh:
```diff
@@ -245,19 +267,32 @@ namespace components
+    /// \return Pointer to a component. Null if the component
+    /// type could not be handled.
+    public: std::unique_ptr<components::BaseComponent> New(
+        const ComponentTypeId &_type, const components::BaseComponent *_data)
```
Can we add some unit tests for this function in `ComponentFactory_TEST`? I'm mainly concerned about cases when the data passed in doesn't match `ComponentTypeId`.

---

As a side note, I'm trying to understand whether the `New` / `Create` functions that receive a raw `BaseComponent` pointer are being added out of necessity or out of convenience. There are all kinds of issues that may happen when mixing types, so we should be extra careful with those new APIs.

adlarkin replied:
Unit tests have been added in https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/d871df3ea4b941bfdec22c2aa8c1df7955a529fa

> As a side note, I'm trying to understand whether the `New` / `Create` functions that receive a raw `BaseComponent` pointer are being added out of necessity or out of convenience

They are added out of necessity (I think :sweat_smile:) I explained why the `Create` function was needed in https://github.com/ignitionrobotics/ign-gazebo/pull/856#discussion_r684550937. The `New` function is used here, for example:

https://github.com/ignitionrobotics/ign-gazebo/blob/660268e70d79899563cf63bdbc114b005a311e32/src/EntityComponentManager.cc#L644-L645

As you can see, we are creating a component and setting its data in a non-templated method. This is possible because `New` has access to the `BaseComponent`'s derived class's `data` via `Create`, which is templated (see https://github.com/ignitionrobotics/ign-gazebo/pull/856#discussion_r684550937).

chapulina replied:
Ok, I think I understand now. It's all because we don't have the component storage classes anymore to create them for us, I believe. Thanks!

chapulina reviewed (commented):
I did another pass, there are only a few places that still concern me and I think deserve more testing. I also didn't go through the list of failed CI tests, but it looks like there are too many?

chapulina commented on include/ignition/gazebo/EntityComponentManager.hh:
```diff
@@ -206,14 +209,16 @@ namespace ignition
       public: template<typename ComponentTypeT>
-              const ComponentTypeT *Component(const ComponentKey &_key) const;
+              const ComponentTypeT *IGN_DEPRECATED(6) Component(
+              const ComponentKey &_key) const;
```
Windows is not happy about something here:

```
C:\Jenkins\workspace\ign_gazebo-pr-win\ws\ign-gazebo\include\ignition/gazebo/EntityComponentManager.hh(212,37): error C2059: syntax error: '' (compiling source file C:\Jenkins\workspace\ign_gazebo-pr-win\ws\ign-gazebo\src\EntityComponentManager.cc) [C:\Jenkins\workspace\ign_gazebo-pr-win\ws\build\ignition-gazebo6\src\ignition-gazebo6.vcxproj]
```

Maybe there's some weird character hiding in there? :woman_shrugging:

adlarkin replied:
It seems like it has something to do with using `IGN_DEPRECATED` in combination with methods that return a pointer. I am guessing that the `*` in the method signature is not being understood correctly when parsing occurs on windows. I removed all of the `IGN_DEPRECATED` macros for methods that return pointers in https://github.com/ignitionrobotics/ign-gazebo/commit/576efc7c5cdae78df4542b07b4249f265d4a6916 (but left the `IGN_DEPRECATED` macros for methods that do not return pointers), and these build errors no longer appear. However, I'm now seeing errors with the function definitions for the templated `View` class (https://build.osrfoundation.org/job/ign_gazebo-pr-win/2858/consoleFull#-11549300628ff58640-3599-4406-a210-216932f1748c):

```
C:\Jenkins\workspace\ign_gazebo-pr-win\ws\ign-gazebo\include\ignition/gazebo/detail/View.hh(150,27): error C2491: 'ignition::gazebo::v6::detail::View<ComponentTypeTs...>::View': definition of dllimport function not allowed (compiling source file C:\Jenkins\workspace\ign_gazebo-pr-win\ws\build\ignition-gazebo6\src\gui\ignition-gazebo6-gui_autogen\mocs_compilation.cpp) [C:\Jenkins\workspace\ign_gazebo-pr-win\ws\build\ignition-gazebo6\src\gui\ignition-gazebo6-gui.vcxproj]
```

@j-rivero do you have any thoughts about what's going on with the `IGN_DEPRECATED` macro, and also how to resolve the `definition of dllimport function not allowed` error above?

adlarkin replied:
Okay, so it sounds like the `IGN_DEPRECATED` issue has been resolved in https://github.com/ignitionrobotics/ign-gazebo/pull/856#issuecomment-896228298. However, I still believe that the `definition of dllimport function not allowed` error remains. Does anyone have any thoughts about how to handle this other build error?

adlarkin commented on include/ignition/gazebo/components/Factory.hh:
```diff
@@ -63,11 +69,21 @@ namespace components
+    public: virtual std::unique_ptr<BaseComponent> Create(
+                const components::BaseComponent *_data) const override
+    {
+      ComponentTypeT comp(*static_cast<const ComponentTypeT *>(_data));
```
> Can't the caller themselves directly create a unique pointer with the component, instead of creating a raw pointer and calling this function to copy it into a unique pointer?

Technically, yes, but this is where the method is used: https://github.com/ignitionrobotics/ign-gazebo/blob/f0a0fe85029fe448b1c12ca522db8ce6f67dc71d/include/ignition/gazebo/components/Factory.hh#L270-L286

`Factory::New` is not a templated method, so I believe that we need to create the unique pointer in `ComponentDescriptor::Create` in order to create an instance of the derived component class, since the `ComponentDescriptor` class is a template class where the class' template parameter is the derived component class. Does that make sense?

chapulina replied:
Ahh gotcha, makes sense. Thanks.

adlarkin commented on include/ignition/gazebo/detail/BaseView.hh:
```diff
@@ -0,0 +1,214 @@
+using ComponentTypeKey = std::vector<ComponentTypeId>;
+
+/// \brief Hash functor for ComponentTypeKey
+struct ComponentTypeHasher
```
> Can you add some comments here about the rationale for the chosen implementation? Maybe some external references?

Here are some references for the hash function I chose:
* https://stackoverflow.com/questions/20511347/a-good-hash-function-for-a-vector
* https://stackoverflow.com/questions/4948780/magic-number-in-boosthash-combine
* https://stackoverflow.com/questions/35985960/c-why-is-boosthash-combine-the-best-way-to-combine-hash-values
* https://softwareengineering.stackexchange.com/questions/402542/where-do-magic-hashing-constants-like-0x9e3779b9-and-0x9e3779b1-come-from

If anyone has thoughts for a better approach, I'd be open to hearing it. Otherwise, if this approach seems fine, I can add these links as documentation.

> It would also be interesting to write a simple unit test for this function that hashes a few component types and checks that they don't collide.

I can add this test coverage :+1: I'll do that once we settle on a hash function.

chapulina replied:
> if this approach seems fine

I don't have a proposal for a better hash function

adlarkin replied:
Tests and references have been added in https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/d871df3ea4b941bfdec22c2aa8c1df7955a529fa

adlarkin commented on include/ignition/gazebo/detail/EntityComponentManager.hh:
```diff
@@ -81,11 +83,24 @@ auto CompareData = [](const DataType &_a, const DataType &_b) -> bool
+        << ". This should never happen!\n";
+      return comp;
+    }
+    *comp = _data;
```
> Isn't `CreateComponentImplementation` already setting the data?

It's not :frowning_face: since the data is part of the derived component class, it can't be set in `CreateComponentImplementation` because `CreateComponentImplementation` only has access to the `BaseComponent` class. So, we have to update the data in the template method that calls `CreateComponentImplementation`, since the template method has access to the derived class and its data. There's a comment that explains this: https://github.com/ignitionrobotics/ign-gazebo/blob/f0a0fe85029fe448b1c12ca522db8ce6f67dc71d/src/EntityComponentManager.cc#L671-L674

adlarkin commented:
> I also didn't go through the list of failed CI tests, but it looks like there are too many?

I'm looking at the Ubuntu CI failures, and I cannot reproduce them on my machine locally (all tests pass locally for me). However, it looks like these failures started in https://build.osrfoundation.org/job/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/6727/, which is the build for #943. So, I'm guessing that one of the changes made there is causing these failures? Looking at the failed build I linked, it says that it had to do with commit https://github.com/ignitionrobotics/ign-gazebo/pull/943/commits/76ffa42c2ba3cfca6a681f16b6aa21c9ee474e3e, which actually appears to be a merge commit that received updates from one of the recent forward ports. But I'm not sure why I can't reproduce these failures locally :thinking:

chapulina commented:
@adlarkin , placing the `*` for raw pointers _after_ `IGN_DEPRECATED` seems to satisfy Windows:

https://github.com/ignitionrobotics/ign-sensors/pull/90/commits/fdde936e1aadd7755117dbbf94bc777340d3a948

adlarkin commented on include/ignition/gazebo/components/Factory.hh:
```diff
@@ -245,19 +267,32 @@ namespace components
+    public: std::unique_ptr<components::BaseComponent> New(
+        const ComponentTypeId &_type, const components::BaseComponent *_data)
+    {
+      // Create a new component if a FactoryFn has been assigned to this type.
```
removed in https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/d871df3ea4b941bfdec22c2aa8c1df7955a529fa

adlarkin commented on include/ignition/gazebo/detail/EntityComponentManager.hh:
```diff
@@ -167,23 +182,27 @@ bool EntityComponentManager::SetComponentData(const Entity _entity,
-  return comp->SetData(_data, CompareData<typename ComponentTypeT::Type>);
+  auto changed = comp->SetData(_data,
+      CompareData<typename ComponentTypeT::Type>);
+  return changed;
```
fixed in https://github.com/ignitionrobotics/ign-gazebo/pull/856/commits/d871df3ea4b941bfdec22c2aa8c1df7955a529fa

chapulina commented:
> Ubuntu CI failures

I also can't reproduce them locally. Looking through the logs, I think it may be due to root permissions on the Jenkins node? That would be related to https://github.com/ignition-tooling/release-tools/pull/494. 

```
27: [ RUN      ] ServerRepeat/ServerFixture.DefaultServerConfig/0
27: [Wrn] [Filesystem.cc:376] Failed to create file [/root/.ignition/gazebo/server.config]: No such file or directory
27: [Err] [ServerConfig.cc:881] Failed to copy installed config [/usr/share/ignition/ignition-gazebo6/server.config] to default config [/root/.ignition/gazebo/server.config].
```

chapulina reviewed (approved):
Looking great! Let's run another round of performance tests to sanity check that we didn't introduce any regressions, and also figure out what's happening with CI.

adlarkin commented:
> Let's run another round of performance tests to sanity check that we didn't introduce any regressions

I ran performance tests again and did not see any noticeable changes in the numbers. You can see the results in the PR description.

> figure out what's happening with CI.

I think at this point, there are 2 lingering CI issues:
1. Test failures on ubuntu CI that cannot be reproduced locally - see https://github.com/ignitionrobotics/ign-gazebo/pull/856#issuecomment-896351597
2. Compilation failure on windows related to the new templated view class - see https://github.com/ignitionrobotics/ign-gazebo/pull/856#discussion_r686238225

chapulina commented:
> Compilation failure on windows 

I'm blindly trying @j-rivero's approach from https://github.com/ignitionrobotics/ign-gui/commit/5f7f05cb21438ddd1564567380b5c8c528061a3d in 66bb2a2

It's worth noting that it fails when compiling the `ignition-gazebo6-gui` component, not the core library. 

> Test failures on ubuntu CI 
>> I think it may be due to root permissions on the Jenkins node?

Nope, the same happens on GitHub actions, so it's definitely an issue with the PR. Very strange that it fails to create the `server.config` file exactly for those tests, but succeeds on other tests. I thought it may be some interference between the tests, but I can't reproduce the failure even running all `colcon test`s locally.

adlarkin commented:
> Nope, the same happens on GitHub actions, so it's definitely an issue with the PR

That's strange - I mentioned in a [previous comment](https://github.com/ignitionrobotics/ign-gazebo/pull/856#issuecomment-895625301) that tests started failing in https://build.osrfoundation.org/job/ignition_gazebo-ci-pr_any-ubuntu_auto-amd64/6727/, which was a build for https://github.com/ignitionrobotics/ign-gazebo/pull/943. But, I don't see how changes in #943 would cause a break.

chapulina commented:
I have a promising fix for `Server_TEST` in d9a49e6, I was able to reproduce the failure locally when the `~/.ignition/gazebo` directory didn't exist. It still doesn't make sense to me why this manifested in this PR though :woman_shrugging:

adlarkin commented:
> I was able to reproduce the failure locally when the ~/.ignition/gazebo directory didn't exist.

Ohhh you know what, this has actually happened to me before. I just forgot about it and always ran `ign gazebo` once (to populate `~/.ignition/gazebo`) before running tests to avoid this error locally. So, if anything, it sounds like this PR exposed a bug instead of creating a new one :wink: Hopefully your fix works! :crossed_fingers:

iche033 reviewed (commented):
I tested this branch with various worlds, spawning and deleting objects, distributed sim, custom plugins and it seems to be working quite well.

the latest ubuntu build failed due to network issue. I just queued another one.

chapulina commented:
That's a lot of failing tests on Ubuntu. The sensors ones should be fixed by #617 , but that's only 9 tests and there are 23 failing here

chapulina commented:
> there are 23 failing here

I'm seeing 23 failing tests on Ubuntu Jenkins for multiple PRs. I believe this is the fallout from https://github.com/ignition-tooling/release-tools/pull/494. 

Let's rely on the other platforms to merge this PR. GitHub actions was happy, as well as homebrew and Windows. I've fixed conflicts and will merge if CI comes back happy for those platforms again.
